### PR TITLE
[mlir] Add support for parsing nested PassPipelineOptions

### PR DESCRIPTION
**Pull request by nikalra**
- Added a default parsing implementation to `PassOptions` to allow `Option`/`ListOption` to wrap PassOption objects. This is helpful when creating meta-pipelines (pass pipelines composed of pass pipelines).
- Updated `ListOption` printing to enable round-tripping the output of `dump-pass-pipeline` back into `mlir-opt` for more complex structures.

### CODE REVIEW

File: mlir/include/mlir/Pass/PassOptions.h [hunk 1]
```diff
@@ -160,13 +179,19 @@ class PassOptions : protected llvm::cl::SubCommand {
   /// TODO: We should upstream the methods in GenericOptionParser to avoid the
   /// need to do this.
```
Comment (River707) [inline mlir/include/mlir/Pass/PassOptions.h:180]: This TODO will need to shift, it's attached to the original conditional

File: mlir/lib/Pass/PassRegistry.cpp [hunk 2]
```diff
@@ -159,13 +160,36 @@ const PassPipelineInfo *mlir::PassPipelineInfo::lookup(StringRef pipelineArg) {
...
+}
+
 LogicalResult detail::pass_options::parseCommaSeparatedList(
     llvm::cl::Option &opt, StringRef argName, StringRef optionStr,
     function_ref<LogicalResult(StringRef)> elementParseFn) {
   // Functor used for finding a character in a string, and skipping over
   // various "range" characters.
   llvm::unique_function<size_t(StringRef, size_t, char)> findChar =
-      [&](StringRef str, size_t index, char c) -> size_t {
+      [&findChar](StringRef str, size_t index, char c) -> size_t {
```
Comment (River707) [inline mlir/lib/Pass/PassRegistry.cpp:192]: Why did this change?

File: mlir/lib/Pass/PassRegistry.cpp [hunk 3]
```diff
@@ -187,13 +211,15 @@ LogicalResult detail::pass_options::parseCommaSeparatedList(
   size_t nextElePos = findChar(optionStr, 0, ',');
   while (nextElePos != StringRef::npos) {
     // Process the portion before the comma.
-    if (failed(elementParseFn(optionStr.substr(0, nextElePos))))
+    if (failed(
+            elementParseFn(extractArgAndUpdateOptions(optionStr, nextElePos))))
       return failure();
 
-    optionStr = optionStr.substr(nextElePos + 1);
+    optionStr = optionStr.drop_front(); // drop the leading ','
```
Comment (River707) [inline mlir/lib/Pass/PassRegistry.cpp:218]:
```suggestion
    // Drop the leading ','.
    optionStr = optionStr.drop_front();
```

Prefer non-trailing comments.

File: mlir/lib/Pass/PassRegistry.cpp [hunk 4]
```diff
@@ -159,13 +160,36 @@ const PassPipelineInfo *mlir::PassPipelineInfo::lookup(StringRef pipelineArg) {
...
+    for (const auto &escape : escapePairs) {
+      if (str.front() == escape.first && str.back() == escape.second) {
+        // Drop the escape characters and trim.
+        str = str.drop_front().drop_back().trim();
+        // Don't process additional escape sequences.
+        break;
+      }
+    }
+  }
+  return str;
```
Comment (River707) [inline mlir/lib/Pass/PassRegistry.cpp:183]:
```suggestion
  if (str.size() <= 2)
    return str;
    
  // Handle escape sequences
  const auto escapePairs = {std::make_pair('\'', '\''),
                            std::make_pair('"', '"'),
                            std::make_pair('{', '}')};
  for (const auto &escape : escapePairs) {
    if (str.front() == escape.first && str.back() == escape.second) {
      // Drop the escape characters and trim.
      str = str.drop_front().drop_back().trim();
      // Don't process additional escape sequences.
      break;
    }
  }
  return str;
```

I know this is the same as before, but can you flip to early exit?

File: mlir/test/lib/Pass/TestPassManager.cpp [hunk 5]
```diff
@@ -66,7 +66,15 @@ struct TestOptionsPass
...
-                         clEnumValN(1, "one", "Example one value"))};
+                         clEnumValN(1, "one", "Example one value"),
+                         clEnumValN(2, "two", "Example two value"))};
+
+    Options() = default;
+    Options(const Options &rhs) { *this = rhs; }
+    Options &operator=(const Options &rhs) {
+      copyOptionValuesFrom(rhs);
+      return *this;
+    }
```
Comment (River707) [inline mlir/test/lib/Pass/TestPassManager.cpp:77]: Can you add these to the PassPipelineOptions<T> base class?

**Review by River707 — APPROVED** [review]
Nice, thanks!

File: mlir/lib/Pass/PassRegistry.cpp (hunk 2 above)
Comment (nikalra) [reply mlir/lib/Pass/PassRegistry.cpp]: Reverted -- I spent a couple of minutes re-reading this lambda to convince myself it didn't implicitly capture/modify anything, but it doesn't impact this change so I'll change it back.

File: mlir/test/lib/Pass/TestPassManager.cpp (hunk 5 above)
Comment (nikalra) [reply mlir/test/lib/Pass/TestPassManager.cpp:77]: ListOption has an implicitly deleted copy constructor so we're forced to add this boilerplate to the derived classes.

**Comment by nikalra** [comment]
@River707 Once everything looks good to you, would you mind merging this PR in? I don't have merge permissions for the LLVM repo.

**Pull request merged**